### PR TITLE
[Infra] collect dumps of all corerun in tree on timeout

### LINKED ISSUE #38643: [Infra] Collect dumps of all CoreRun exes in the process tree on test timeout
I've noticed that timeouts for tests that leverage things like `RemoteExecutor` or custom remoting infrastructure will result in a dump only for the _first_ instance of `corerun` under the xunit test runner.

I propose that we update the dumping logic to dump every instance of `corerun` under that initial instance as well on timeout.  This would collect dumps for the "remote" process as well as the originating process.  I have a patch that I haven't tested yet that makes this change.  This issue is meant to track that work/proposal.

CC @jashook @jkoritzinsky

### PR DESCRIPTION
**Pull request by josalem**
* adds `ProcessExtensions` for centralizing process tree traversal
* gets all `corerun` exes under the test runner
* use `createdump` for all plats

I'll be adding a mock test that times out in order to test this across our matrix of platforms.  Until that is removed, I'll be marking this as no-merge.

resolves #38643

CC @tommcdon @jkoritzinsky @jashook

### CODE REVIEW

**Comment by danmosemsft** [comment]
Will this work for libraries tests or just clr tests?

**Comment by jkoritzinsky** [comment]
This is just for the clr tests (in particular the CoreCLR test runs, I don't believe this works with Mono). The libraries tests require switching to `dotnet test` and using their built-in dump collection on timeout support as far as I know. @ViktorHofer would know if there's a way to enable dumps for the libraries tests before we switch to `dotnet test`.

**Comment by josalem** [comment]
> Will this work for libraries tests or just clr tests?

Like Jeremy said, these changes will only work for CLR tests.  I used the wrong area tag when I created the PR I think.  That being said, I don't think there is anything unique to the CLR tests in these changes and the code should be reusable elsewhere.

File: src/coreclr/tests/src/Common/Coreclr.TestWrapper/CoreclrTestWrapperLib.cs
Comment (jkoritzinsky) [inline src/coreclr/tests/src/Common/Coreclr.TestWrapper/CoreclrTestWrapperLib.cs:241]: Since you're already rewriting this completely, it may be worth trying to switch to `createdump.exe` so we're consistent across platforms (and get more usage of `createdump.exe`).

**Comment by jkoritzinsky** [comment]
Since this is checking for CoreRun processes, it will miss the native COM tests that have a native entry point and start up the runtime via CoreShim.dll. It would be nice (but I wouldn't consider it blocking) if this could be extended to check for processes that loaded CoreShim.dll when on Windows.

**Comment by josalem** [comment]
> Since this is checking for CoreRun processes, it will miss the native COM tests that have a native entry point and start up the runtime via CoreShim.dll. It would be nice (but I wouldn't consider it blocking) if this could be extended to check for processes that loaded CoreShim.dll when on Windows.

I think there are other tests that have entrypoints/binaries not named `corerun`, e.g., Profiler tests.  Depending on the processes the test runner spins up, we could just dump all ancestor processes of the first `corerun` under the test wrapper.  Once I've got the basic functionality working, I can try that approach.

**Comment by jkoritzinsky** [comment]
In the CoreShim case, the process would be a child of the corerun process (we use a managed program to run the native program because it was the easiest with the current test harness setup)

**Comment by josalem** [comment]
@mikem8361, it looks like `createdump` is failing to dump the lowest subprocess (the grandchild of the test process), but succeeding on the child of the test process (the first instance of `corerun) for non-Windows platforms.  Any idea why?  It happens consistently, so I'm guessing this is something you can get under a debugger locally.  `createdump` appears to just die mid-dump and doesn't print anything to `stderr` or `stdout` about what happened.

Other than the `createdump` issues with the youngest descendent process, it looks like the logic for collecting the child processes _is_ working as seen in the Windows legs.  I'll start cleaning up the diagnostics printing and get this ready for review.

These changes should also fix collecting dumps entirely on ARM Ubuntu runs, since it looks like the ARM builds of Ubuntu don't have the necessary kernel configuration to use the `/proc/pid/task/pid/children` pseudo-file.  The fallback to `pgrep` works in those cases.

**Comment by josalem** [comment]
I'm leaving the purposefully failing test and by extension the NO-MERGE label till after review so behavior can be observed, then I'll remove it before merging.

**Comment by mikem8361** [comment]
I filed issue https://github.com/dotnet/runtime/issues/38905 to track the createdump/DAC enum memory segfault when the runtime is paused on start in early initialization.

**Comment by josalem** [comment]
Here's a sample of what the output will look like:
image

and a link to the console output with multiple dumps taken: https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-38647-merge-132ddbfd45114f9f85/PayloadGroup0/console.f794bddf.log?sv=2019-02-02&se=2020-07-30T01%3A24%3A45Z&sr=c&sp=rl&sig=wC2yIEkMHKWS2rhYP%2FkmNga0cSPgzDdj4iPeQ%2F8LI9c%3D

File: src/tests/Common/Coreclr.TestWrapper/CoreclrTestWrapperLib.cs
Comment (hoyosjs) [inline src/tests/Common/Coreclr.TestWrapper/CoreclrTestWrapperLib.cs:233]: 20 min to dump a process sounds a little excessive. Even if it took that long to take a dump, uploading and downloading/loading it would be rather impractical. `WaitForExit` also blocks the thread which if we have a bad deadlock we'll see hangs in CI with lots of threads creating dumps. I haven't tried createdump on macOS to give you a good estimate of a good timeout here (lldb is pretty awful..., don't know how it compares).

File: src/tests/Common/Coreclr.TestWrapper/CoreclrTestWrapperLib.cs
Comment (josalem) [reply src/tests/Common/Coreclr.TestWrapper/CoreclrTestWrapperLib.cs:99]: I just moved this code around.  I'll take a look at the docs and add an error case for that.

File: src/tests/Common/Coreclr.TestWrapper/CoreclrTestWrapperLib.cs
Comment (josalem) [reply src/tests/Common/Coreclr.TestWrapper/CoreclrTestWrapperLib.cs:200]: There definitely need to be spaces there 😝 .  I'll add units as well.  @jashook, I know you've been looking at timing data in CI, do you have any thoughts on modifying the default timeout here?  I'm going to leave it as is for now.

File: src/tests/Common/Coreclr.TestWrapper/CoreclrTestWrapperLib.cs
Comment (josalem) [reply src/tests/Common/Coreclr.TestWrapper/CoreclrTestWrapperLib.cs:233]: Ah, the `* 2` was only added when I was testing the functionality and trying to pinpoint some timing issues.  I doubt it would take 20 minutes in _most_ cases for coreclr pri0 tests.  The original timeout was `DEFAULT_TIMEOUT` (10 minutes), which still seems extreme.  I'll put it back to 10 minutes, but we might want to revisit whether that still makes sense.

**Comment by josalem** [comment]
I'm going to remove the purposefully failing test since the behavior appears to be working across the supported matrix.  Once that is done I will remove the "NO MERGE" label.

Here is a link to the AzDO test results for the build that shows 2 dumps for the timed out test across all OS x Arch combos: https://dev.azure.com/dnceng/public/_build/results?buildId=723752&view=ms.vss-test-web.build-test-results-tab.

image

File: src/tests/Common/Coreclr.TestWrapper/CoreclrTestWrapperLib.cs
Comment (hoyosjs) [reply src/tests/Common/Coreclr.TestWrapper/CoreclrTestWrapperLib.cs:233]: This is still a bit scary to me for dumping timeouts.

Comment (josalem) [reply src/tests/Common/Coreclr.TestWrapper/CoreclrTestWrapperLib.cs:233]: > `WaitForExit` also blocks the thread which if we have a bad deadlock we'll see hangs in CI with lots of threads creating dumps.

I assume you're talking about `createdump` deadlocking here?  We could follow the timeout up with a call to `Process.Kill()` in the event of timeout to make sure we don't leave any zombied `createdump` processes in CI.  I'm guessing the the file IO is going to be the limiting factor when it comes to how long it takes for `createdump` to run to completion.

Comment (josalem) [reply src/tests/Common/Coreclr.TestWrapper/CoreclrTestWrapperLib.cs:233]: Hmmm... Now that I'm thinking about it... I don't think timed-out tests are killed either.  Let me add that and run one more CI with the purposefully failing test.

File: src/coreclr/src/debug/createdump/crashinfomac.cpp
Comment (josalem) [reply src/coreclr/src/debug/createdump/crashinfomac.cpp:33]: I think this is because he did a `git revert` and then made the same changes in another branch or reused the commits when he merged.  Any idea how to undo this removal?  Is it just a matter of removing his commits from this PR instead of just `revert`ing them?

Comment (hoyosjs) [reply src/coreclr/src/debug/createdump/crashinfomac.cpp:33]: I'd cherry pick them out, yet.

Comment (mikem8361) [reply src/coreclr/src/debug/createdump/crashinfomac.cpp:33]: Sorry about that.

Comment (mikem8361) [reply src/coreclr/src/debug/createdump/crashinfomac.cpp:33]: I meant to remove all of those changes.

Comment (josalem) [reply src/coreclr/src/debug/createdump/crashinfomac.cpp]: `git rebase -i` helped me remove the commits entirely 👍

**Review by hoyosjs — APPROVED** [review]

**Comment by josalem** [comment]
@jashook, who all should review this before we merge the changes to infra?

**Review by jkoritzinsky — APPROVED** [review]

File: src/tests/Common/Coreclr.TestWrapper/CoreclrTestWrapperLib.cs
Comment (jashook) [inline src/tests/Common/Coreclr.TestWrapper/CoreclrTestWrapperLib.cs:221]: Is sudo absolutely required here? I believe this currently just "happens" to work because we have root access in the linux containers

Comment (jashook) [reply src/tests/Common/Coreclr.TestWrapper/CoreclrTestWrapperLib.cs:221]: Keep in mind that our containers are also not explicitly required to have sudo installed as well, they may just run as su, because of this it seems fragile to use it in this way.

Comment (jashook) [reply src/tests/Common/Coreclr.TestWrapper/CoreclrTestWrapperLib.cs:221]: Also I understand that this was existing code. This is mostly a question going forward

File: src/tests/Common/Coreclr.TestWrapper/CoreclrTestWrapperLib.cs
Comment (jashook) [inline src/tests/Common/Coreclr.TestWrapper/CoreclrTestWrapperLib.cs:219]: Did we verify we have access to sudo on OSX

**Review by jashook — COMMENTED** [review]
lgtm minus two questions

File: src/tests/Common/Coreclr.TestWrapper/CoreclrTestWrapperLib.cs
Comment (josalem) [reply src/tests/Common/Coreclr.TestWrapper/CoreclrTestWrapperLib.cs:219]: I ran a couple runs with a purposefully timing out test to validate that this logic works, and it collected multiple dumps for each platform.  I believe that it is working across the board. @mikem8361 would know more of any limitations of createdump on mac.

File: src/tests/Common/Coreclr.TestWrapper/CoreclrTestWrapperLib.cs
Comment (josalem) [reply src/tests/Common/Coreclr.TestWrapper/CoreclrTestWrapperLib.cs:221]: I'm not sure for Linux.  If we are already running as `root` I don't think it is necessary.  As for macOS, however, I think the permissions in 10.15+ require `sudo` for createdump to work.  @mikem8361 ?

**Comment by josalem** [comment]
I think this change is going to be critical for current and future tracing CI investigations, so I'll be merging this PR this morning if there aren't any objections.

**Pull request merged**
